### PR TITLE
[WIP] Add copy functionality

### PR DESCRIPTION
**Pull request by amyrlam**
~~Has [LiveReload WIP commits](https://github.com/amyrlam/Docusaurus/pull/2) for ease of development~~

### CODE REVIEW

**Comment by yangshun** [comment]
This is awesome! It works at the moment, but as you said, some styling improvements need to be made. I have some suggestions for structural improvements as well.

Since we want to ship this to all Docusaurus users, it should be part of the core folder, within lib. This PR is currently adding the functionality to Docusaurus' website only, not the Docusaurus library. Try shifting the code into the library, within `lib`. We can create a new directory called `scripts` or  and add the `code-block-buttons.js` file in there instead. Also, we will need to tweak the server-side code to serve that file during development and generate.

Try to separate the LiveReload commits out as we want to merge this PR independently from LiveReload.

File: website/siteConfig.js [hunk 1]
```diff
@@ -180,7 +180,11 @@ const siteConfig = {
   highlight: {
     theme: 'atom-one-dark',
   },
-  scripts: ['https://buttons.github.io/buttons.js'],
+  scripts: [
+    'https://buttons.github.io/buttons.js',
+    'https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.0/clipboard.min.js',
```
Comment (yangshun) [inline website/siteConfig.js:192]: These two scripts have to be shifted into either the default `siteConfig` or into the HTML of the base body structure. Putting it here only affects Docusaurus' website, not its users.

File: website/static/css/custom.css [hunk 2]
```diff
@@ -5,3 +5,47 @@
...
+.btn-icon svg {
+  fill: currentColor;
+  margin-right: .4em;
+}
+
+.btn-icon__label {
+  font-size: 11px;
+}
+
+.btn-open-in-repl {
```
Comment (yangshun) [inline website/static/css/custom.css]: Not needed

File: website/static/css/custom.css [hunk 3]
```diff
@@ -5,3 +5,47 @@
 table td:first-child > code {
   white-space: nowrap;
 }
+
```
Comment (yangshun) [inline website/static/css/custom.css]: These CSS should be part of `main.css` as well so as to affect Docusaurus users.

File: website/static/js/code-blocks-buttons.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  addButtons(
+    ".hljs",
+    button("Copy", "Copy code to clipboard", copyIcon, "btn-clipboard")
+  );
+
+  const tryIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="12" width="12" viewBox="0 0 64 64"><path fill="currentColor" d="M32 64C14.4 64 0 49.6 0 32S14.4 0 32 0s32 14.4 32 32-14.4 32-32 32zm0-58.7C17.3 5.3 5.3 17.3 5.3 32s12 26.7 26.7 26.7 26.7-12 26.7-26.7S46.7 5.3 32 5.3z"/><path fill="currentColor" d="M42.1 31.1l-15.9-9.4c-1.2-.7-2.1-.2-2.1 1.2v18.3c0 1.4 1 1.9 2.1 1.2L42.1 33s.6-.4.6-.9c0-.6-.6-1-.6-1z"/></svg>';
```
Comment (yangshun) [inline website/static/js/code-blocks-buttons.js]: This is for Babel website's playground button, and is not needed here.

File: website/static/js/code-blocks-buttons.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    event.clearSelection();
+    const textEl = event.trigger.querySelector(".btn-icon__label");
+    textEl.textContent = "Copied";
+    setTimeout(function() {
+      textEl.textContent = "Copy";
+    }, 2000);
+  });
+
+  // not sure what this is
+  const presets = ["es2015", "stage-0", "react"];
```
Comment (yangshun) [inline website/static/js/code-blocks-buttons.js]: This are babel presets, also used only for the playground.

**Review by yangshun — COMMENTED** [review]
Please remove the code that is used only for Babel Playground. Try to understand what the code is doing and remove where unnecessary.

File: website/siteConfig.js (hunk 1 above)
Comment (amyrlam) [reply website/siteConfig.js:192]: That makes sense. I kept these here and copied them to `examples/basics/siteConfig.js`, following buttons.github.io. Is that what you meant?

File: website/static/css/custom.css (hunk 2 above)
Comment (amyrlam) [reply website/static/css/custom.css]: Fixed, sorry I shouldn't have left all of this in, just wanted to post what I had functional at EOD.

File: website/static/css/custom.css (hunk 3 above)
Comment (amyrlam) [reply website/static/css/custom.css]: Fixed, I moved this to `lib/static/css/main.css`.

File: website/static/js/code-blocks-buttons.js (hunk 4 above)
Comment (amyrlam) [reply website/static/js/code-blocks-buttons.js]: 👍

File: website/static/js/code-blocks-buttons.js (hunk 5 above)
Comment (amyrlam) [reply website/static/js/code-blocks-buttons.js]: 👍

**Comment by amyrlam** [comment]
Removed the LiveReload stuff.

//

(1)
Should the directory for `code-block-buttons.js` be `static/js` or `scripts`? 

(2)
> Also, we will need to tweak the server-side code to serve that file during development and generate.

It seems to work with no changes? Confused why `js/code-blocks-buttons` works before and after [this commit](https://github.com/amyrlam/Docusaurus/pull/3/commits/b8210ea2396401eeda9adf926388a9a5988903bd)?

//

Let me know, can make more changes tomorrow.

//

As for the styling itself, now it's within the code block, no color change on highlight. 

I think it would look better if the blue outline on click is flush with the edges of the code block. But figured I'd just show this to you first to see if even a viable design. Since I suppose I would just measure that with "guess and check" in the browser.

copy

**Comment by amyrlam** [comment]
I think this is ready for Docusaurus? I'm going to open a new PR on the main repo. I'll make any needed edits there. Sorry this one was so noisy.

**Comment by yangshun** [comment]
Ok let's submit a PR and see. I just added preview builds last night. Excited to see it in action 😄 

Also test `docusaurus-init` locally and make sure it works for a fresh init of docusaurus.

**Comment by amyrlam** [comment]
Ah ok. I did a fresh `docusaurus-init` in Docusaurus repo and it's not working. Debugging it...

Also, the color scheme to "match" with the dark code block on Docusaurus' blog rn will not be a good default

copy

**Comment by amyrlam** [comment]
@yangshun Oh I guess that within Docusaurus, I'm still pinned to 1.0.14 Docusaurus version. 

On this branch I did a `rm -rf website`. Instead of https://github.com/facebook/Docusaurus/blob/master/website/package.json#L16, I'm linked to     `"docusaurus": "^1.0.14"`.

I manually changed that to `github:amyrlam/Docusaurus#amy/add-copy"`.

I then add 
```
```js
headerLinks: [
    ...
    { blog: true, label: 'Blog' },
    ...
]```
``` 
to a blog post, copy does not appear.

So I'm not sure what's wrong yet. Will further debug tomorrow. Maybe I have some code in the wrong folder, where it's not a part of the `docusaurus-init` process?

Is this the right way to test this locally? 

//

The default color of "Copy" will also have to be changed, so that it works with the Docusaurus default.

Our color scheme:

copy-functionality

Default(?) color scheme (from docusaurus-starter-kit):

image

Maybe just change to black.

**Comment by amyrlam** [comment]
moving PR to main site with questions

**Pull request closed without merging**